# Week 1 Assignment 1: Tokenization Analysis

## Objective
Analyze how GPT-4's tokenizer handles a non-English text compared to its English translation, quantifying the "multilingual tax" and identifying tokenizer bias.

## Key Concepts
- **Tokenization**: Process of breaking text into tokens (units processed by the model)
- **Multilingual Tax**: The inefficiency of tokenizing non-English text (more tokens needed for same content)
- **Tokenizer Bias**: Systematic differences in how the tokenizer splits words in different languages

## Step 1: Select a Source Text

**Task**: Choose a non-English paragraph (100-150 words)

**Source**: Wikipedia article on 2024

**My chosen language**: German

**Original German text** (110 words):
```
2024 war geprägt durch die im Jahresverlauf fortgesetzten Kriege, dem Russisch-Ukrainischen Krieg und dem Krieg in Israel und Gaza. In Deutschland zerbrach nach einer gescheiterten Vertrauensfrage die Regierung von Bundeskanzler Scholz. Mit erstmals mehr als 1,5 Grad Celsius über dem vorindustriellen Mittel war das Jahr 2024 das heißeste seit Beginn der Aufzeichnungen. Weltweit kam es zu extremen Wetterereignissen wie Dürre, Überschwemmungen und Waldbränden. Die gesellschaftlichen Debatten drehten sich besonders um Klimaschutz, Migration und neue Sicherheitsgesetze. In vielen Ländern kam es zu Parlaments- und Präsidentschaftswahlen, die mitunter zu erheblichen politischen Veränderungen führten. Die wirtschaftliche Entwicklung war weiterhin von Unsicherheiten geprägt, unter anderem durch Energiekrisen und Lieferkettenprobleme. Zugleich wurden bedeutende Fortschritte in der Digitalisierung und beim Einsatz künstlicher Intelligenz erzielt.
```

## Step 2: Prepare the English Version

**Task**: Find or create an English translation of the same text.

**English translation** (106 words):
```
2024 was characterized by the wars that continued throughout the year, the Russian-Ukrainian war and the war in Israel and Gaza. In Germany, Chancellor Scholz's government collapsed after a failed vote of confidence. For the first time with more than 1.5 degrees Celsius above the pre-industrial average, 2024 was the hottest year since records began. Worldwide, extreme weather events such as droughts, floods and wildfires occurred. Public debates centered particularly on climate protection, migration and new security laws. In many countries, parliamentary and presidential elections took place, which sometimes led to significant political changes. Economic development continued to be characterized by uncertainties, including energy crises and supply chain problems. At the same time, significant progress was made in digitalization and the use of artificial intelligence.
```

In [1]:
# Store both texts for analysis

german_text = """2024 war geprägt durch die im Jahresverlauf fortgesetzten Kriege, dem Russisch-Ukrainischen Krieg und dem Krieg in Israel und Gaza. In Deutschland zerbrach nach einer gescheiterten Vertrauensfrage die Regierung von Bundeskanzler Scholz. Mit erstmals mehr als 1,5 Grad Celsius über dem vorindustriellen Mittel war das Jahr 2024 das heißeste seit Beginn der Aufzeichnungen. Weltweit kam es zu extremen Wetterereignissen wie Dürre, Überschwemmungen und Waldbränden. Die gesellschaftlichen Debatten drehten sich besonders um Klimaschutz, Migration und neue Sicherheitsgesetze. In vielen Ländern kam es zu Parlaments- und Präsidentschaftswahlen, die mitunter zu erheblichen politischen Veränderungen führten. Die wirtschaftliche Entwicklung war weiterhin von Unsicherheiten geprägt, unter anderem durch Energiekrisen und Lieferkettenprobleme. Zugleich wurden bedeutende Fortschritte in der Digitalisierung und beim Einsatz künstlicher Intelligenz erzielt."""

english_text = """2024 was characterized by the wars that continued throughout the year, the Russian-Ukrainian war and the war in Israel and Gaza. In Germany, Chancellor Scholz's government collapsed after a failed vote of confidence. For the first time with more than 1.5 degrees Celsius above the pre-industrial average, 2024 was the hottest year since records began. Worldwide, extreme weather events such as droughts, floods and wildfires occurred. Public debates centered particularly on climate protection, migration and new security laws. In many countries, parliamentary and presidential elections took place, which sometimes led to significant political changes. Economic development continued to be characterized by uncertainties, including energy crises and supply chain problems. At the same time, significant progress was made in digitalization and the use of artificial intelligence."""

print(f"German text length: {len(german_text)} characters")
print(f"English text length: {len(english_text)} characters")
print(f"\nGerman word count: {len(german_text.split())} words")
print(f"English word count: {len(english_text.split())} words")

German text length: 950 characters
English text length: 879 characters

German word count: 118 words
English word count: 125 words


## Step 3: Tokenize Using GPT-4 Tokenizer

**Task**: Use the Tokenizer demo to tokenize both texts.

**Results from tokenizer demo:**
- German text: **259 tokens** (GPT-4 / GPT-3.5)
- English text: **160 tokens** (GPT-4 / GPT-3.5)

[Screenshots provided in submission]

## Step 4: Record Token Counts

**Tokenization Results from GPT-4 Tokenizer**:

In [2]:
# Token counts from the tokenizer demo

german_tokens = 259
english_tokens = 160

print(f"✓ German text tokens: {german_tokens}")
print(f"✓ English text tokens: {english_tokens}")
print(f"\nDifference: {german_tokens - english_tokens} extra tokens in German")
print(f"German text is {(german_tokens/english_tokens - 1)*100:.1f}% longer in tokens")

✓ German text tokens: 259
✓ English text tokens: 160

Difference: 99 extra tokens in German
German text is 61.9% longer in tokens


## Step 5: Calculate the Multilingual Tax

**Formula**: Multilingual Tax = (German Tokens - English Tokens) / English Tokens × 100

In [3]:
# Calculate the multilingual tax

multilingual_tax = ((german_tokens - english_tokens) / english_tokens) * 100

print(f"\n" + "="*70)
print(f"🚨 MULTILINGUAL TAX: {multilingual_tax:.2f}%")
print(f"="*70)

print(f"\nWhat this means:")
print(f"  • German requires {german_tokens - english_tokens} more tokens than English")
print(f"  • For every 100 tokens in English, German needs {100 + multilingual_tax:.0f} tokens")
print(f"  • German users pay {multilingual_tax:.2f}% MORE for the exact same content")
print(f"  • German users get {multilingual_tax:.2f}% LESS usable context window space")


🚨 MULTILINGUAL TAX: 61.88%

What this means:
  • German requires 99 more tokens than English
  • For every 100 tokens in English, German needs 162 tokens
  • German users pay 61.88% MORE for the exact same content
  • German users get 61.88% LESS usable context window space


## Step 6: Identify Tokenizer Bias

**Task**: Find 3 specific examples where the tokenizer splits German words differently than English.

### Example 1: Lieferkettenprobleme (Supply Chain Problems)

**German word**: Lieferkettenprobleme (one compound word)

**How tokenizer splits it**: Lie | fer | ket | ten | pro | ble | me = **7 tokens**

**English equivalent**: supply chain problems (three separate words)

**How tokenizer splits it**: supply | chain | problems = **3 tokens**

**Why this shows bias**: 
German naturally forms compound words, which the tokenizer doesn't recognize as a unit. The same concept in English uses three familiar, common words that each become single tokens. The tokenizer's vocabulary was trained primarily on English, so it has dedicated tokens for "supply," "chain," and "problems," but not for German compound components. This forces the German word to be broken into sub-word pieces, while English gets efficient word-level tokenization.

---

### Example 2: Energiekrisen (Energy Crises)

**German word**: Energiekrisen (one compound word)

**How tokenizer splits it**: E | nergie | k | ris | en = **5 tokens**

**English equivalent**: energy crises (two separate words)

**How tokenizer splits it**: energy | crises = **2 tokens**

**Why this shows bias**: 
Again, a German compound word that semantically represents one concept gets fragmented into 5 pieces, while the English equivalent stays as 2 clean tokens. The tokenizer has "energy" and "crises" as vocabulary items but doesn't have efficient tokenization for German compound formation. This is because the training data heavily favored English.

---

### Example 3: Wetterereignissen (Weather Events)

**German word**: Wetterereignissen (one compound word, plural dative)

**How tokenizer splits it**: W | etter | ere | ign | issen = **5 tokens**

**English equivalent**: weather events (two separate words)

**How tokenizer splits it**: weather | events = **2 tokens**

**Why this shows bias**: 
Similar pattern: German compounds get fragmented (5 tokens) while English words stay intact (2 tokens). The tokenizer even splits "Wetterereignissen" into awkward boundaries (W | etter instead of Wetter), showing the vocabulary doesn't have German morphological units. This reveals the tokenizer was optimized for English structure, not German compound formation or case/number inflection.

In [4]:
# Document your bias findings

bias_examples = [
    {
        "german_word": "Lieferkettenprobleme",
        "german_tokens": ["Lie", "fer", "ket", "ten", "pro", "ble", "me"],
        "german_token_count": 7,
        "english_equivalent": "supply chain problems",
        "english_tokens": ["supply", "chain", "problems"],
        "english_token_count": 3,
        "analysis": "German compounds are tokenized as subword pieces while English words stay intact. The tokenizer has vocabulary for English words but not for German compound components."
    },
    {
        "german_word": "Energiekrisen",
        "german_tokens": ["E", "nergie", "k", "ris", "en"],
        "german_token_count": 5,
        "english_equivalent": "energy crises",
        "english_tokens": ["energy", "crises"],
        "english_token_count": 2,
        "analysis": "Another German compound broken into 5 tokens vs 2 English tokens. Shows systematic bias against German word formation patterns."
    },
    {
        "german_word": "Wetterereignissen",
        "german_tokens": ["W", "etter", "ere", "ign", "issen"],
        "german_token_count": 5,
        "english_equivalent": "weather events",
        "english_tokens": ["weather", "events"],
        "english_token_count": 2,
        "analysis": "Awkward tokenization boundaries (even splitting 'Wetter' unnaturally) demonstrate the tokenizer wasn't trained to recognize German morphological units or inflections."
    }
]

print("TOKENIZER BIAS EXAMPLES")
print("=" * 70)
for i, example in enumerate(bias_examples, 1):
    print(f"\nExample {i}: {example['german_word']}")
    print(f"  German ({example['german_token_count']} tokens): {' | '.join(example['german_tokens'])}")
    print(f"  English ({example['english_token_count']} tokens): {' | '.join(example['english_tokens'])}")
    print(f"  Ratio: German uses {example['german_token_count']/example['english_token_count']:.1f}x more tokens")
    print(f"  Analysis: {example['analysis']}")

TOKENIZER BIAS EXAMPLES

Example 1: Lieferkettenprobleme
  German (7 tokens): Lie | fer | ket | ten | pro | ble | me
  English (3 tokens): supply | chain | problems
  Ratio: German uses 2.3x more tokens
  Analysis: German compounds are tokenized as subword pieces while English words stay intact. The tokenizer has vocabulary for English words but not for German compound components.

Example 2: Energiekrisen
  German (5 tokens): E | nergie | k | ris | en
  English (2 tokens): energy | crises
  Ratio: German uses 2.5x more tokens
  Analysis: Another German compound broken into 5 tokens vs 2 English tokens. Shows systematic bias against German word formation patterns.

Example 3: Wetterereignissen
  German (5 tokens): W | etter | ere | ign | issen
  English (2 tokens): weather | events
  Ratio: German uses 2.5x more tokens
  Analysis: Awkward tokenization boundaries (even splitting 'Wetter' unnaturally) demonstrate the tokenizer wasn't trained to recognize German morphological units or inf

## Step 7: Understanding the Practical Implications

### 1. Cost Implications
- **You pay per token**: GPT-4 costs ~$0.03 per 1,000 tokens
- **Multilingual tax = direct cost increase**: 61.88% more tokens = 61.88% higher costs
- **Scale matters**: At enterprise scale, this becomes significant

### 2. Context Window Implications
- **Limited context**: GPT-4 has an 8K or 128K token context window
- **Less usable space for non-English**: 62% fewer tokens available means 62% fewer documents fit
- **Feature limitations**: Long document analysis, summarization, retrieval tasks become harder in German

### 3. Design and User Experience Implications
- **Fairness issue**: Non-English speakers pay more for worse performance
- **Product decisions**: Features that work in English may not scale to German
- **Localization cost**: Must budget extra for non-English versions

In [5]:
# Calculate practical implications

print("="*70)
print("PRACTICAL IMPLICATIONS OF TOKENIZATION BIAS")
print("="*70)

print(f"\n1️⃣  COST IMPLICATIONS")
print(f"   ─────────────────────────────────────")
extra_tokens = german_tokens - english_tokens
cost_per_1k = 0.03

cost_per_request_english = (english_tokens * cost_per_1k) / 1000
cost_per_request_german = (german_tokens * cost_per_1k) / 1000
extra_cost_per_request = cost_per_request_german - cost_per_request_english

print(f"   Extra tokens per request: {extra_tokens}")
print(f"   Cost per request (English): ${cost_per_request_english:.4f}")
print(f"   Cost per request (German): ${cost_per_request_german:.4f}")
print(f"   Extra cost per request: ${extra_cost_per_request:.4f}")
print(f"   ")
print(f"   Per 100,000 requests: ${extra_cost_per_request * 100_000:.2f} extra")
print(f"   Per 1,000,000 requests: ${extra_cost_per_request * 1_000_000:.2f} extra")
print(f"   Annual (10M requests): ${extra_cost_per_request * 10_000_000:.2f} extra")

print(f"\n2️⃣  CONTEXT WINDOW IMPLICATIONS")
print(f"   ─────────────────────────────────────")
context_window_8k = 8000
context_window_128k = 128000

english_paragraphs_8k = context_window_8k // english_tokens
german_paragraphs_8k = context_window_8k // german_tokens
lost_capacity_8k = english_paragraphs_8k - german_paragraphs_8k

english_paragraphs_128k = context_window_128k // english_tokens
german_paragraphs_128k = context_window_128k // german_tokens
lost_capacity_128k = english_paragraphs_128k - german_paragraphs_128k

print(f"   With 8K context window:")
print(f"     • English: {english_paragraphs_8k} paragraphs fit")
print(f"     • German: {german_paragraphs_8k} paragraphs fit")
print(f"     • Lost capacity: {lost_capacity_8k} fewer paragraphs for German")

print(f"\n   With 128K context window:")
print(f"     • English: {english_paragraphs_128k} paragraphs fit")
print(f"     • German: {german_paragraphs_128k} paragraphs fit")
print(f"     • Lost capacity: {lost_capacity_128k} fewer paragraphs for German")

print(f"\n3️⃣  USER EXPERIENCE IMPACT")
print(f"   ─────────────────────────────────────")
print(f"   Multilingual Tax: {multilingual_tax:.2f}%")
print(f"   ")
print(f"   For a German user:")
print(f"   • They pay {multilingual_tax:.1f}% more per API call than English speakers")
print(f"   • They have {multilingual_tax:.1f}% less context window space")
print(f"   • This affects EVERY application using GPT-4 in German")
print(f"   • Over time, this compounds as a fairness and accessibility issue")

PRACTICAL IMPLICATIONS OF TOKENIZATION BIAS

1️⃣  COST IMPLICATIONS
   ─────────────────────────────────────
   Extra tokens per request: 99
   Cost per request (English): $0.0048
   Cost per request (German): $0.0078
   Extra cost per request: $0.0030
   
   Per 100,000 requests: $297.00 extra
   Per 1,000,000 requests: $2970.00 extra
   Annual (10M requests): $29700.00 extra

2️⃣  CONTEXT WINDOW IMPLICATIONS
   ─────────────────────────────────────
   With 8K context window:
     • English: 50 paragraphs fit
     • German: 30 paragraphs fit
     • Lost capacity: 20 fewer paragraphs for German

   With 128K context window:
     • English: 800 paragraphs fit
     • German: 494 paragraphs fit
     • Lost capacity: 306 fewer paragraphs for German

3️⃣  USER EXPERIENCE IMPACT
   ─────────────────────────────────────
   Multilingual Tax: 61.88%
   
   For a German user:
   • They pay 61.9% more per API call than English speakers
   • They have 61.9% less context window space
   • This affe

## Step 8: Write Your Analysis

### YOUR ANALYSIS

**Part 1: Tokenization Efficiency Results**

My analysis reveals a significant multilingual tax of 61.88% for German compared to English. The German text required 259 tokens while the English translation used only 160 tokens—a difference of 99 extra tokens for the same semantic content. This means German-speaking users would pay 61.88% more per API request when using GPT-4. This finding was surprising in its magnitude; while I expected German to be less efficient due to compound word formation, I did not anticipate such a large gap. The nearly 2.5x token multiplier suggests systemic bias in the tokenizer's design rather than just linguistic differences.

**Part 2: Tokenizer Bias Evidence**

The three examples I identified clearly demonstrate English-centric tokenizer bias. First, "Lieferkettenprobleme" (supply chain problems) fragments into 7 tokens (Lie|fer|ket|ten|pro|ble|me) while the English equivalent uses only 3 tokens (supply|chain|problems). Similarly, "Energiekrisen" (energy crises) breaks into 5 tokens (E|nergie|k|ris|en) versus 2 English tokens (energy|crises). Finally, "Wetterereignissen" (weather events) becomes 5 awkwardly-split tokens (W|etter|ere|ign|issen) compared to 2 English tokens (weather|events). The pattern is clear: German compound words are fragmented into subword pieces while English maintains word-level boundaries. This reveals that the tokenizer's vocabulary was trained primarily on English text, giving it efficient encoding for common English words but forcing it to use character-level or subword approximations for German.

**Part 3: Design Implications**

If I were building a product for both English and German users, I would need to make several design trade-offs. First, I'd need to acknowledge this limitation transparently—German users deserve to know they're paying more for the same functionality. For features like document summarization or long-form analysis, I might need to offer reduced document lengths or implement multi-step processing for German users. For cost modeling, I'd factor in the 62% premium and either absorb that cost difference or implement tiered pricing that's fair across languages. Most importantly, I would advocate for better tokenizers that treat all languages equitably. At the product level, I'd prioritize features that work well within German's tokenization constraints—shorter, more focused queries rather than long documents—and design the UX around these realities rather than pretending the problem doesn't exist.

## Screenshots & Evidence

### Screenshot 1: German Text Tokenization
German: 259 tokens (GPT-4 / GPT-3.5)

### Screenshot 2: English Text Tokenization  
English: 160 tokens (GPT-4 / GPT-3.5)

[Screenshots provided in submission]

## Summary & Key Takeaways

| Metric | Result |
|--------|--------|
| **German tokens** | 259 |
| **English tokens** | 160 |
| **Token difference** | 99 tokens |
| **Multilingual Tax** | 61.88% |
| **Key bias pattern** | German compounds fragment into 3-7 tokens while English stays at 1-3 tokens |
| **Main design implication** | Non-English users pay significantly more and have reduced functionality; transparency and fair pricing are essential |

### Reflection Questions:

1. **Why does this matter?** 
   Tokenization bias creates systemic inequity for non-English users. It's not just a technical quirk—it affects real costs and capabilities.

2. **What's the root cause?** 
   GPT-4's tokenizer was trained primarily on English-heavy datasets, optimizing for English word patterns and vocabulary at the expense of other languages.

3. **How should products handle this?** 
   Acknowledge the limitation transparently, build it into cost models, and design features that work within language-specific constraints.

4. **What's the future?** 
   Better multilingual tokenizers that give equal weight to major languages could significantly reduce this gap and make AI more equitable globally.